### ISSUE TITLE
vllm prepends two BOS for LLama

### ISSUE THREAD

**Issue by wenquanlu** [huggingface/trl#3853]
### Reproduction

While I was working with grpo_trainer.py and Llama-3.2-3B-Instruct, I noticed a significant drop in reward if using vllm compared to vanilla transformers inference. I did a deep investigation, and found that vllm prepends an extra bos token during generation, as shown below

```python
                with profiling_context(self, "vLLM.generate"):
                    all_outputs = self.llm.generate(all_prompts_text, sampling_params=sampling_params, use_tqdm=False)
                    print("vllm token id", all_outputs[0].prompt_token_ids)

```

outputs (I am using GRPO vllm collocate mode): 

```
vllm token id [128000, 128000, 128006, 9125, 128007, 271, 38766, 1303, 33025, 2696, 25, 6790, 220, 2366, 18, 198, 15724, 2696, 25, 220, 2705, 5033, 220, 2366, 20, 271, 2675, 527, 264, 11190, 15592, 22103, 430, 5825, 1664, 5621, 1525, 291, 323, 11944, 14847, 13, 1472, 1176, 1781, 922, 279, 33811, 1920, 439, 459, 5419, 1647, 77828, 323, 1243, 3493, 279, 1217, 449, 279, 4320, 13, 40633, 304, 279, 2768, 3645, 25, 366, 27963, 397, 9522, 524, 27963, 397, 27, 9399, 397, 9522, 524, 9399, 29, 128009, 128006, 882, 128007, 271, 25797, 2106, 323, 35266, 527, 3515, 264, 14130, 311, 1518, 889, 649, 636, 279, 1888, 12239, 304, 279, 538, 13, 2684, 617, 1027, 220, 24, 32272, 323, 41769, 706, 264, 220, 5925, 304, 279, 538, 13, 35266, 706, 264, 220, 6083, 13, 578, 1620, 16720, 374, 5922, 279, 1890, 3392, 439, 682, 279, 1023, 32272, 13, 35266, 2751, 264, 220, 1954, 389, 279, 1620, 16720, 13, 3639, 374, 279, 8187, 12239, 41769, 3966, 311, 636, 311, 9567, 35266, 422, 682, 28711, 527, 4459, 5219, 30, 128009, 128006, 78191, 128007, 271]
    ...
```

You can see there are two 128000 at the start. I think this issue was mentioned in here https://github.com/vllm-project/vllm/issues/9519 

but seems nothing get addressed.

### System Info

trl: 0.18.0
python: 3.11.13
transformers: 4.52.3
vllm: 0.8.5.post1

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by pramodith** [comment]
@wenquanlu I tried the following script

with 

vllm == 0.9.1
transformers==4.53.3
trl==0.22.0

```
with tempfile.TemporaryDirectory() as tmp_dir:
    training_args = GRPOConfig(
        output_dir=tmp_dir,
        learning_rate=0.1,  # increase the learning rate to speed up the test
        per_device_train_batch_size=4,  # reduce the batch size to reduce memory usage
        num_generations=4,  # reduce the number of generations to reduce memory usage
        max_completion_length=8,  # reduce the completion length to reduce memory usage
        report_to="none",
        use_vllm=True,
        vllm_mode="colocate",
        vllm_tensor_parallel_size=1,
        vllm_gpu_memory_utilization=0.5,
        max_prompt_length=512,
        top_p=0.9,
        top_k=10,
        min_p=0.01,
        repetition_penalty=1.1,
    )

    trainer = GRPOTrainer(
        model="FuseAI/FuseChat-Llama-3.2-1B-Instruct",  # tiny model is too small for vLLM
        reward_funcs=[reward_func],
        args=training_args,
        train_dataset=dataset,
    )
    trainer.train()
```


Inside grpo_trainer.py I updated.

```
with profiling_context(self, "vLLM.generate"):
        all_outputs = self.llm.generate(vllm_inputs, sampling_params=sampling_params, use_tqdm=False)
        print(f"Outputs: {all_outputs[0].prompt_token_ids}")  # Debugging output to check the generated completions
```

I don't see duplicated bos_tokens in the prompt though. Wonder if you just need to update your dep versions?

**Comment by wenquanlu** [comment]
Hi @pramodith , can you try using meta-llama/Llama-3.2-1B-Instruct ? thanks

also, the open-r1 README instruct us to use vllm==0.8.5.post1, maybe they need to update their instructions

**Comment by pramodith** [comment]
@wenquanlu unfortunately, I can't. For some reason my request to use the llama-3.2 family of models has been denied by Meta 😅 . I think the model I used `FuseAI/FuseChat-Llama-3.2-1B-Instruct` is a finetuned version of Llama-3.2 and when I look at the tokenizers [config](https://huggingface.co/FuseAI/FuseChat-Llama-3.2-1B-Instruct/blob/main/special_tokens_map.json) it seems to have the same special tokens, so the behavior of the tokenizer shouldn't be different.

**Comment by wenquanlu** [comment]
Hi @pramodith , thanks for your reply! are you able to replicate the issue under my setting ?
trl: 0.18.0
python: 3.11.13
transformers: 4.52.3
vllm: 0.8.5.post1

because this is the official environment listed in open-r1, https://github.com/huggingface/open-r1/blob/main/setup.py

**Comment by pramodith** [comment]
@wenquanlu I'm running into a bunch of dep issues on my linux vm when I try to match your settings.

**Comment by albertvillanova** [comment]
Hey @wenquanlu, sorry for the very late reply — we're catching up on the backlog. 

Your diagnosis was spot on: in TRL 0.18.0 the GRPO trainer passed the decoded prompt *text* to vLLM's `generate()`, and since vLLM re-tokenizes with `add_special_tokens=True` by default, it added a second BOS on top of the one already inserted by the chat template — hence the `[128000, 128000, ...]` you saw. 

This was fixed in #5238, which tokenizes the prompt up front and hands raw `prompt_token_ids` to vLLM, so vLLM no longer re-tokenizes and the double BOS can't occur (see https://github.com/huggingface/trl/blob/b77f36f3bb9558b9c319fb17af1a37833e8d1e90/trl/generation/vllm_generation.py#L705-L716). 

Closing as fixed — please re-open if you still hit this on a current TRL release.
